File: dev/serve.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""開発者モード用の簡易配信サーバ(Python 標準ライブラリのみ / npm・Node 不要)。

dist/ を CORS 付きで配信する。SP ページ(https)から http://127.0.0.1 への
fetch はブラウザの localhost 例外で許可される。
全リクエストを1行ログに出す(届いているかの切り分け用 — 全アプリ共通規約 §18)。

使い方:  python3 dev/serve.py [port]   (既定 18086、環境変数 PERMREG_DEV_PORT でも指定可)
"""
import os
import sys
from datetime import datetime
from functools import partial
from http.server import SimpleHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path

PORT = int(os.environ.get('PERMREG_DEV_PORT') or (sys.argv[1] if len(sys.argv) > 1 else 18086))
DIST = Path(__file__).resolve().parent.parent / 'dist'


class Handler(SimpleHTTPRequestHandler):
    def translate_path(self, path):
        # /permreg/xxx も /xxx も dist/xxx に解決する
        clean = path.split('?')[0]
        if clean.startswith('/permreg/') or clean == '/permreg':
            clean = clean[len('/permreg'):] or '/'
        self.path = clean
        return super().translate_path(clean)

    def end_headers(self):
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Cache-Control', 'no-store')
        super().end_headers()

    def log_message(self, fmt, *args):
        origin = self.headers.get('Origin', '') if hasattr(self, 'headers') else ''
        print(datetime.now().strftime('%H:%M:%S'), fmt % args, origin)


if __name__ == '__main__':
    handler = partial(Handler, directory=str(DIST))
    print(f'permreg dev server: http://127.0.0.1:{PORT}/permreg/ -> dist/')
    ThreadingHTTPServer(('127.0.0.1', PORT), handler).serve_forever()

File: dev/serve.js
// 開発者モード用の簡易配信サーバ: dist/ を CORS 付きで配信する。
// SP ページ(https)から http://127.0.0.1 への fetch はブラウザの localhost 例外で許可される。
// 全リクエストを1行ログに出す(届いているかの切り分け用 — 全アプリ共通規約 §18)。
const http = require('http');
const fs = require('fs');
const path = require('path');

const PORT = +(process.env.PERMREG_DEV_PORT || process.argv[2] || 18086);
const DIST = path.join(__dirname, '..', 'dist');
const TYPES = { '.js': 'text/javascript; charset=utf-8', '.txt': 'text/plain; charset=utf-8', '.html': 'text/html; charset=utf-8' };
const now = () => new Date().toISOString().slice(11, 19);

http.createServer((req, res) => {
  const headers = { 'Access-Control-Allow-Origin': '*', 'Cache-Control': 'no-store' };
  // /permreg/xxx も /xxx も dist/xxx に解決する
  const rel = decodeURIComponent((req.url || '/').split('?')[0]).replace(/^\/permreg/, '') || '/';
  const file = path.normalize(path.join(DIST, rel));
  if (!file.startsWith(DIST)) {
    res.writeHead(403, headers);
    res.end('forbidden');
    return;
  }
  fs.readFile(file, (err, data) => {
    const status = err ? 404 : 200;
    console.log(now(), req.method, req.url, status, req.headers.origin || '');
    if (err) {
      res.writeHead(404, headers);
      res.end('not found');
      return;
    }
    res.writeHead(200, Object.assign({ 'Content-Type': TYPES[path.extname(file)] || 'application/octet-stream' }, headers));
    res.end(data);
  });
}).listen(PORT, '127.0.0.1', () => {
  console.log('permreg dev server: http://127.0.0.1:' + PORT + '/permreg/ -> dist/');
});
